### DISCUSSION TITLE
How to use MAX31865 and PT1000 ?

### DISCUSSION THREAD

**Discussion by Noschvie** [arendst/Tasmota#14420] [Q&A] [answered]
Hello
want to us a PT1000 (2 wire) with a MAX31865 board connected to a Wemos D1 mini.
CS -> D8 / 15
SDI -> D7 / 13
SDO -> D6 / 12
CLK -> D5 / 14
GND -> GND
Vin -> 3V3

Tasmota (latest dev) self compiled with:

```
#define USE_SPI                                  // Hardware SPI using GPIO12(MISO), GPIO13(MOSI) and GPIO14(CLK) in addition to two user selectable GPIOs(CS and DC)

#define USE_MAX31865                           // Add support for MAX31865 RTD sensors using softSPI
#define MAX31865_PTD_WIRES  2                  // PTDs come in several flavors, pick yours. Specific settings per sensor possible with MAX31865_PTD_WIRES1..MAX31865_PTD_WIRES6
#define MAX31865_PTD_RES    1000               // Nominal PTD resistance at 0°C (100Ω for a PT100, 1000Ω for a PT1000, YMMV!). Specific settings per sensor possible with MAX31865_PTD_RES1..MAX31865_PTD_RES6
#define MAX31865_REF_RES    4300               // Reference resistor (Usually 430Ω for a PT100, 4300Ω for a PT1000). Specific settings per sensor possible with MAX31865_REF_RES1..MAX31865_REF_RES6
#define MAX31865_PTD_BIAS   0                  // To calibrate your not-so-good PTD. Specific settings per sensor possible with MAX31865_PTD_BIAS1..MAX31865_PTD_BIAS6
```

But if the MAX31865 is connected to 3V3 Tasmota doesn't startup / running.

Is this a valid configuration for Tasmota ?

The Arduino MAX31865  example is running on this hardware setup and provides valid temperature values.

Any idea ? Thanks !

**Comment by lalo-uy** [comment]
D8 needs a pulldown in orther to let esp to boot.
Try 1k to gnd.

**Reply by barbudor** [reply]
Not necessary as the Wemos D1 already has that on board

Are you 100% sure that the Arduino sample code is working with the exact same setup (wires, etc...) and you haven't touched anything between Arduino sample and Tasmota ?

**Reply by Noschvie** [reply]
Yes, have the same setup running the tests.

running under Windows, Arduino 1.8.13
[my Arduino Example: Adafruit P1000 RTD Sensor / MAX31865](https://gist.github.com/Noschvie/2340489bb96f4c8767546f80ab736191#file-test-max31865-ino)

Steps
.) disconnect 3V3 between Wemos and MAX31865
.) Upload the sketch to Wemos
.) open serial monitor
.) connect 3V3 between Wemos and MAX31865

with this output
```
07:37:19.082 -> RTD value: 8315
07:37:19.082 -> Ratio = 0.25375366
07:37:19.082 -> Resistance = 1091.14074707
07:37:19.082 -> Temperature = 23.40
07:37:19.128 -> 
07:37:21.232 -> RTD value: 8317
07:37:21.232 -> Ratio = 0.25381470
07:37:21.232 -> Resistance = 1091.40319824
07:37:21.232 -> Temperature = 23.43
07:37:21.285 -> 
07:37:23.389 -> RTD value: 8317
07:37:23.389 -> Ratio = 0.25381470
07:37:23.389 -> Resistance = 1091.40319824
07:37:23.389 -> Temperature = 23.47
07:37:23.436 -> 
07:37:25.494 -> RTD value: 8319
07:37:25.494 -> Ratio = 0.25387573
07:37:25.494 -> Resistance = 1091.66564941
07:37:25.540 -> Temperature = 23.57
```

**Reply by barbudor** [reply]
I see a possible difference in your test protocol as you are connecting the 3V3 to the MAX after upload and reset of the board (openning serial terminal is resetting the board).
So it's likely the ESP has already booted when you connect the MAX.

With Tasmota you must have the MAX connected from the very beginning to have it detected by Tasmota in it's boot sequence. If you connect it after and Tasmota will not detect it.
So when you connect it before booting the ESP, then the ESP doesn't boot.

It is likely your module is preventing the ESP to boot, presumably because it is pulling GPIO15 up.
There is already a 10K pull down resistor on the Wemos D1 but if the MAX module as a stronger pull-up (ie a smaller resistance value), this makes a voltage divider. If voltage on GPIO15 is higher than ~0.5V, the ESP doesn't see that anymore as a LOW signal and that can is the reason for the boot to fail.

Please look carefully at your MAX module for a resistor between the CS line and 3V3
2 options there:
1) try another GPIO for CS which will not have a problem with a pull-up (likely any other GPIO, avoid however GPIO16/D0)
2) identify and remove the problematic resistor to have it working on GPIO15. After Tasmota has booted, Tasmota will configure this pin as an output so pull-up is not needed.

**Comment by Noschvie** [comment]
Changed my setup to use GPIO02/D4 for CS / SSPI CS.

Now using Arduino works with conected 3V3, but Tasmota doesn't.

Maybe it's a problem with the self-compiled Tasmota (the environment works for SDM230 Device).

Arduino uses software SPI, isn't it? And Tasmota too, isn't it?
Is `#define USE_SPI` needed to be included ?

It's the same behaviour using a "NodeMCU Amica Modul V2 ESP8266 ESP-12F".

Thanks!

**Reply by barbudor** [reply]
Please elaborate on "Tasmota doesn't work"

Is it the same problem (no boot) ?
Or a different problem like booting but not displaying or displaying wrong values ?

**Reply by Noschvie** [reply]
Sorry... Tasmota is booting but not displaying values.
Please find below the configuration and the log of the boot sequence.
Is this a valid configuration ?

![grafik](https://user-images.githubusercontent.com/48217146/149570677-2bde477d-e78f-4840-aadb-054f941bf471.png)
```
00:00:00.001 HDW: ESP8266EX
00:00:00.050 CFG: aus Flash geladen von FB, Anzahl 10
00:00:00.115 QPC: Reset
00:00:00.120 SPI: Software using GPIO14(CLK), GPIO13(MOSI) and GPIO12(MISO)
00:00:00.126 Projekt tasmota - Tasmota Version 2022.01.2(tasmota)-2_7_4_9(2022.01.13 16:24:26)
00:00:00.515 WIF: verbinden mit AP1 Gentzgasse_5G Channel 6 BSSId 34:FC:B9:B7:14:80 in Modus 11n als tasmota-CEA829-2089...
00:00:01.752 WIF: verbunden
00:00:02.005 HTP: Web-Server aktiv bei tasmota-CEA829-2089 mit IP-Adresse 192.168.1.169
20:02:10.019 MQT: Verbindungsversuch...
20:02:10.031 MQT: verbunden
20:02:10.034 MQT: tele/tasmota_CEA829/LWT = Online (beibehalten)
20:02:10.037 MQT: cmnd/tasmota_CEA829/POWER = 
20:02:10.043 MQT: tele/tasmota_CEA829/INFO1 = {"Info1":{"Module":"Generic","Version":"2022.01.2(tasmota)","FallbackTopic":"cmnd/DVES_CEA829_fb/","GroupTopic":"cmnd/tasmotas/"}}
20:02:10.052 MQT: tele/tasmota_CEA829/INFO2 = {"Info2":{"WebServerMode":"Admin","Hostname":"tasmota-CEA829-2089","IPAddress":"192.168.1.169"}}
20:02:10.064 MQT: tele/tasmota_CEA829/INFO3 = {"Info3":{"RestartReason":"Software/System restart"}}
20:02:14.431 MQT: tele/tasmota_CEA829/STATE = {"Time":"2022-01-14T20:02:14","Uptime":"0T00:00:09","UptimeSec":9,"Heap":29,"SleepMode":"Dynamic","Sleep":50,"LoadAvg":19,"MqttCount":1,"Wifi":{"AP":1,"SSId":"Gentzgasse_5G","BSSId":"34:FC:B9:B7:14:80","Channel":6,"Mode":"11n","RSSI":100,"Signal":-45,"LinkCount":1,"Downtime":"0T00:00:03"}}
20:02:22.898 CMD: status 0
20:02:22.905 MQT: stat/tasmota_CEA829/STATUS = {"Status":{"Module":0,"DeviceName":"Tasmota","FriendlyName":["Tasmota"],"Topic":"tasmota_CEA829","ButtonTopic":"0","Power":0,"PowerOnState":3,"LedState":1,"LedMask":"FFFF","SaveData":1,"SaveState":1,"SwitchTopic":"0","SwitchMode":[0,0,0,0,0,0,0,0],"ButtonRetain":0,"SwitchRetain":0,"SensorRetain":0,"PowerRetain":0,"InfoRetain":0,"StateRetain":0}}
20:02:22.934 MQT: stat/tasmota_CEA829/STATUS1 = {"StatusPRM":{"Baudrate":115200,"SerialConfig":"8N1","GroupTopic":"tasmotas","OtaUrl":"http://ota.tasmota.com/tasmota/release/tasmota.bin.gz","RestartReason":"Software/System restart","Uptime":"0T00:00:17","StartupUTC":"2022-01-14T19:02:05","Sleep":50,"CfgHolder":4617,"BootCount":4,"BCResetTime":"2022-01-14T09:29:19","SaveCount":11,"SaveAddress":"FA000"}}
20:02:22.970 MQT: stat/tasmota_CEA829/STATUS2 = {"StatusFWR":{"Version":"2022.01.2(tasmota)","BuildDateTime":"2022.01.13 16:24:26","Boot":31,"Core":"2_7_4_9","SDK":"2.2.2-dev(38a443e)","CpuFrequency":80,"Hardware":"ESP8266EX","CR":"390/699"}}
20:02:22.992 MQT: stat/tasmota_CEA829/STATUS3 = {"StatusLOG":{"SerialLog":2,"WebLog":2,"MqttLog":0,"SysLog":0,"LogHost":"","LogPort":514,"SSId":["Gentzgasse_5G",""],"TelePeriod":300,"Resolution":"558180C0","SetOption":["00008009","2805C80001000600003C5A0A000000000000","00000080","00006000","00004000"]}}
20:02:23.001 MQT: stat/tasmota_CEA829/STATUS4 = {"StatusMEM":{"ProgramSize":622,"Free":380,"Heap":25,"ProgramFlashSize":4096,"FlashSize":4096,"FlashChipId":"1640D8","FlashFrequency":40,"FlashMode":3,"Features":["00000407","8FDAC78F","04368001","000000CF","110013C0","C000F981","00004004","00001000","00000020"],"Drivers":"1,2,3,4,5,6,7,8,9,10,12,16,18,19,20,21,22,24,26,27,29,30,35,37,45","Sensors":"1,2,3,4,5,6,47"}}
20:02:23.033 MQT: stat/tasmota_CEA829/STATUS5 = {"StatusNET":{"Hostname":"tasmota-CEA829-2089","IPAddress":"192.168.1.169","Gateway":"192.168.1.1","Subnetmask":"255.255.255.0","DNSServer1":"192.168.1.1","DNSServer2":"0.0.0.0","Mac":"A8:48:FA:CE:A8:29","Webserver":2,"HTTP_API":1,"WifiConfig":4,"WifiPower":17.0}}
20:02:23.060 MQT: stat/tasmota_CEA829/STATUS6 = {"StatusMQT":{"MqttHost":"192.168.1.43","MqttPort":1883,"MqttClientMask":"DVES_%06X","MqttClient":"DVES_CEA829","MqttUser":"Top24IoTDeviceUser","MqttCount":1,"MAX_PACKET_SIZE":1200,"KEEPALIVE":30,"SOCKET_TIMEOUT":4}}
20:02:23.087 MQT: stat/tasmota_CEA829/STATUS7 = {"StatusTIM":{"UTC":"2022-01-14T19:02:23","Local":"2022-01-14T20:02:23","StartDST":"2022-03-27T02:00:00","EndDST":"2022-10-30T03:00:00","Timezone":99,"Sunrise":"08:39","Sunset":"17:19"}}
20:02:23.107 MQT: stat/tasmota_CEA829/STATUS10 = {"StatusSNS":{"Time":"2022-01-14T20:02:23"}}
20:02:23.116 MQT: stat/tasmota_CEA829/STATUS11 = {"StatusSTS":{"Time":"2022-01-14T20:02:23","Uptime":"0T00:00:18","UptimeSec":18,"Heap":25,"SleepMode":"Dynamic","Sleep":50,"LoadAvg":19,"MqttCount":1,"Wifi":{"AP":1,"SSId":"Gentzgasse_5G","BSSId":"34:FC:B9:B7:14:80","Channel":6,"Mode":"11n","RSSI":100,"Signal":-40,"LinkCount":1,"Downtime":"0T00:00:03"}}}
```

```
20:06:48.039 CMD: Backlog Template; Module; GPIO 255
20:06:48.121 MQT: stat/tasmota_CEA829/RESULT = {"NAME":"Generic","GPIO":[1,1,1,1,1,1,1,1,1,1,1,1,1,1],"FLAG":0,"BASE":18}
20:06:48.336 MQT: stat/tasmota_CEA829/RESULT = {"Module":{"0":"Generic"}}
20:06:48.590 MQT: stat/tasmota_CEA829/RESULT = {"GPIO0":{"0":"None"},"GPIO1":{"0":"None"},"GPIO2":{"928":"SSPI CS"},"GPIO3":{"0":"None"},"GPIO4":{"0":"None"},"GPIO5":{"0":"None"},"GPIO9":{"0":"None"},"GPIO10":{"0":"None"},"GPIO12":{"832":"SSPI MISO"},"GPIO13":{"864":"SSPI MOSI"},"GPIO14":{"896":"SSPI SCLK"},"GPIO15":{"0":"None"},"GPIO16":{"0":"None"},"GPIO17":{"0":"None"}}
```

**Reply by barbudor** [reply]
Your CS pins is wrong, it should "MX31865 CS"

**Comment by Noschvie** [comment, marked as answer]
Works fine!
Thanks and all the best!

![grafik](https://user-images.githubusercontent.com/48217146/149576347-a45a36f4-e4c2-4bc4-8a07-75270e5230b5.png)

![grafik](https://user-images.githubusercontent.com/48217146/149576390-fe4a050c-c6ec-4292-bac8-ec2d63161dae.png)

**Reply by ChristosKeleshis** [reply]
I have trouble reading the temperature value as a number in openhab. Is there any example on how to just retrieve the temp value through mqtt? Thanks

**Reply by Noschvie** [reply]
@ChristosKeleshis find below an example of the sensor message, what is your problem?

`11:35:07.438 MQT: tele/tasmota_85FB6A/SENSOR = {"Time":"1970-02-11T11:35:07","Switch3":"OFF","MAX31865-0":{"Temperature":21.3,"Resistance":1083.0,"Error":0},"TempUnit":"C"}`

**Reply by Jason2866** [reply]
@ChristosKeleshis This is a question for a OpenHab forum. How to parse a mqtt message.

**Comment by johyunsoo** [comment]
Hi. I used ESP32 board and installed the latest version of BIN file. I want to use MAX31865 but I think it only supports MAX31855. The document says it supports MAX31865 but is it actually not supported by ESP32?

**Reply by Noschvie** [reply]
You need to compile Tasmota (self compiled) with the settings from the first post.

**Reply by sfromis** [reply]
As you can see on the pertinent docs page, MAX31865 is not included in precompiled binaries, but MAX31855 is included with ESP32 binaries (or ESP8266 sensors binary).
https://tasmota.github.io/docs/BUILDS/

**Comment by thki00** [comment]
Hi, I'm trying to get my PT1000 running with a ESP8266 (`D1 Mini NodeMcu with ESP8266-12F`) and a MAX31865 (`Hailege PT100 MAX31865 RTD Temperature Thermocouple Sensor Amplifier Module`) and I'm kind of stuck. 

I configured everything the same way as in this discussion. I used the following options to compile Tasmota:
```
#define USE_MAX31865            
#define MAX31865_PTD_WIRES  2              
#define MAX31865_PTD_RES    1000           
#define MAX31865_REF_RES    4300
#define MAX31865_PTD_BIAS   0
```

My wiring:
```
CS -> D4
CLK -> D5
SDO -> D6
SDI -> D7
GND -> G
Vin -> 3V3
```

My Tasmota Setup:
```
D4 GPIO2: MX31865 CS (1)
D6 GPIO12: SSPI MISO
D7 GPIO13: SSPI MOSI
D5 GPIO14: SCLK
```

On the MAX31865 board I soldered together the two triangles with where "2/3 wire" and "2 wire" is written. However, when I connect the PT1000 I always get a temperatur of 988.8 °C


I tired different resistors and this is what I see in the logs then:

// 120 Ohm:
```
21:47:10.427 RSL: SENSOR = {"Time":"2024-02-18T21:47:10","MAX31865-0":{"Temperature":48.3,"Resistance":1187.6,"Error":0},"TempUnit":"C"}
```

// 1200 Ohm:
```
21:43:11.619 RSL: SENSOR = {"Time":"2024-02-18T21:43:11","MAX31865-0":{"Temperature":988.8,"Resistance":4299.9,"Error":128},"TempUnit":"C"}
```

// wires not connected at all:
```
{"Time":"2024-02-19T19:25:25","MAX31865-0":{"Temperature":988.8,"Resistance":4299.9,"Error":128},"TempUnit":"C"}
```

// connected both wires without a resistance:
```
21:45:20.459 RSL: SENSOR = {"Time":"2024-02-18T21:45:20","MAX31865-0":{"Temperature":-241.7,"Resistance":1.3,"Error":0},"TempUnit":"C"}
```

I first thought that tasmota is configured for a PT100, as the 120 Ohm resistance is showing a correct value for a PT100, but I compiled it several times and even replaced the raw value `xsns_47_max31865.ino` files with MAX31865_REF_RES=4300 and MAX31865_PTD_RES=1000, but still it doesn't work.

Does anyone has a hint for me what might be wrong? Do I maybe bought the wrong MAX31865 board that somehow only works with PT100?

**Reply by thki00** [reply]
Okay, I guess I already found out by myself. Looks like my MAX31865  board is only for a PT100, as the Rref has 430 Ohm instead of 4300 Ohm ... I somehow thought that the board can handle both types PT100 / PT1000, but that doesn't seem to be the case.

**Reply by sfromis** [reply]
When doing the custom build, you can select your "reference resistor" value in `user_config_override.h`
https://github.com/arendst/Tasmota/blob/development/tasmota/my_user_config.h#L913
The comment there explicitly mentions both PT100 and PT1000

**Reply by thki00** [reply]
Just tried that and set the value `MAX31865_REF_RES` to 430 and then got this output with a 1200 Ohm resistor:
```
SENSOR = {"Time":"2024-02-19T21:56:56","MAX31865-0":{"Temperature":-142.1,"Resistance":4299.9,"Error":128},"TempUnit":"C"}
```

But should that really work if there is this 430 Rref instead of the 4300 on board? From reading some descriptions / comments I found on Amazon, it seems like the board only supports resistance between 100 and 1000 Ohm, and the PT1000 starts at 1000 ohm (0°C)?

**Reply by sfromis** [reply]
I have no experience with those boards, but I'd try a build with 430 instead of 4300, to check if your thought works out. Ideally, check what reference resistor is physically on your board.

Googling for the title keywords led me to pages suggesting that it has a 430Ω reference resistor. I also saw boards with SMD resistor codes 431 and 4300. Both would mean 430Ω if following standard ways of labeling SMD resistors; the last digit is a multiplier, powers of 10.

**Reply by Noschvie** [reply]
For the PT1000 you have to use this https://amzn.eu/d/ebAKz9M

**Reply by Noschvie** [reply]
I'm using this: 
MAX31865 CS (1) -> D0 / GPIO16
because of D4 / GPIO02 is used for BUILTIN_LED, isn't?

**Comment by Noschvie** [comment]
Have the MAX31865 board running with a PT1000.
These are my settings:

```
#define USE_SPI                                  // Hardware SPI using GPIO12(MISO), GPIO13(MOSI) and GPIO14(CLK) in addition to two user selectable GPIOs(CS and DC)
 
#define USE_MAX31865                           // Add support for MAX31865 RTD sensors using softSPI
#define MAX31865_PTD_WIRES  2                  // PTDs come in several flavors, pick yours. Specific settings per sensor possible with MAX31865_PTD_WIRES1..MAX31865_PTD_WIRES6
#define MAX31865_PTD_RES    1000               // Nominal PTD resistance at 0°C (100Ω for a PT100, 1000Ω for a PT1000, YMMV!). Specific settings per sensor possible with MAX31865_PTD_RES1..MAX31865_PTD_RES6
#define MAX31865_REF_RES    4300               // Reference resistor (Usually 430Ω for a PT100, 4300Ω for a PT1000). Specific settings per sensor possible with MAX31865_REF_RES1..MAX31865_REF_RES6
#define MAX31865_PTD_BIAS   0                  // To calibrate your not-so-good PTD. Specific settings per sensor possible with MAX31865_PTD_BIAS1..MAX31865_PTD_BIAS6
```

**Comment by thki00** [comment]
Thanks for the help! The issue was indeed the wrong MAX31865 board ... I ordered now the "Adafruit PT1000 RTD Temperature Sensor" and that one works fine with the same configuration as I have initially posted.
